### ISSUE TITLE
error on init repo for google cloud storage on restic 11 but works in restic 10

### ISSUE THREAD

**Issue by eleith** [restic/restic#3100]
restic version
--------------------------
restic 0.11.0 compiled with go1.15.3 on linux/amd64                                                      

How did you run restic exactly?
-------------------------------

`restic -r gs:bucket/folder init`

Expected behavior
-----------------

repo is initialized on restic 11.0 and 10.0

Actual behavior
---------------

Fatal: create repository at gs:bucket:/folder failed: service.Buckets.Get: googleapi: Error 403: xyz@cloud-id.iam.gserviceaccount.com does not have storage.buckets.get access to the Google Cloud Storage bucket., forbidden 

but on 10.0 it works

**Comment by fd0** [comment]
Hm, that's odd. We have updated the library to access GCS between 0.10.0 and 0.11.0, so that could be it (in PR #2974). Can you spot any differences which may explain what you're seeing?

From what I can see nothing interesting has changed. Are you able to run a `git bisect` between 0.10.0 and 0.11.0?

**Comment by eleith** [comment]
yup, i think i found in https://github.com/restic/restic/pull/2974 what could be causing it. i'll debug next chance i get and try and confirm the resolution.

**Comment by fd0** [comment]
Nice, looking forward to what you find!

**Comment by eleith** [comment]
for some context, the service account i'm using has `storage.objects.*` permissions but does not have `storage.buckets.get` permission. the reason being that i don't want the service account to be able to create new buckets, just to create objects in an existing bucket (including initializing new restic repositories in the bucket).

in restic 11.0 there is new behavior in how bucket lookup failures are handled. if a bucket can not be looked up (which happens for service accounts that do not have `storage.buckets.get` permission), then it fails to create the repository.

however, in restic 10.0, when a bucket could not be looked up, [it would continue ambiguously forward](https://github.com/restic/restic/pull/2974/files#diff-d575e938da99f8972d0d37a331741e0ad800d9dde75a27a9b096816e128d9595L147) and attempt to create an object in the bucket. if that works, all is well. 

if you wanted to preserve the behavior of restic 10.0, we can modify `Create()` in `internal/backend/gs/gs.go` to the following

```go                                                          
  func Create(cfg Config, rt http.RoundTripper) (restic.Backend, error) {                                
    be, err := open(cfg, rt)                                                                             
    if err != nil {                                                                                      
      return nil, errors.Wrap(err, "open")                                                               
    }                                                                                                    
                                                                                                         
    // Try to determine if the bucket exists. If it does not, try to create it.                          
    ctx := context.Background()                                                                          
    exists, err := be.bucketExists(ctx, be.bucket)                                                       
    if err != nil { 
      // return nil, errors.Wrap(err, "service.Buckets.Get")                                                                                                                            

      /*------------------------------------------------------------------
      +  the service account might still have object permissions
      + just continue forward ambiguously till the next storage call      
      ----------------------------------------------------------------------*/

      return be, nil                                                                                     
    }                                                                                                    
                                                                                                         
    if !exists {                                                                                         
      // Bucket doesn't exist, try to create it.                                                         
      if err := be.bucket.Create(ctx, be.projectID, nil); err != nil {                                   
        // Always an error, as the bucket definitely doesn't exist.                                      
        return nil, errors.Wrap(err, "service.Buckets.Insert")                                           
      }                                                                                                  
                                                                                                         
    }                                                                                                    
                                                                                                         
    return be, nil                                                                                       
  }
```
so in conclusion, there was a behavior change. either the behavior is desired and the documentation could be updated, or the above function could be modified to preserve the behavior.

which would you prefer @fd0 ? 

i wouldn't mind putting together a small PR for this given my current investigation.

**Comment by fd0** [comment]
Cool, thanks for debugging! Would you mind finding out which error is returned by `be.bucketExists()`? You can add the following code within the `if err != nil`:

```go
    exists, err := be.bucketExists(ctx, be.bucket)                                                       
    if err != nil {
        fmt.Fprintf(os.Stderr, "bucketExists() returned error %T %#v\n", err, err)
    }
```

This will print the error's type (via `%T`) and the Go representation of it. Maybe we can check for a specific error instead of just ignoring the error.

**Comment by MichaelEischer** [comment]
It's probably a `googleapi.Error` which has a field `Code` containing the HTTP status. The `ErrBucketNotExist` return value is created in the gcs library by checking that status code against `http.StatusNotFound`. With missing permissions I'd assume the status code to be `http.StatusForbidden` (which was also returned prior to merging #2974)

**Comment by eleith** [comment]
type: `googleapi.Error`
code: `403`

i'll put together a PR catching for this specific error and not blocking. this way, a service account need only object permissions to be able to initialize a repository.

i've also tested trying to create a repository in a new bucket and restic appropriately errors with a message informing the user that the service account needs the `service.Bucket.create` permissions.

**Comment by eleith** [comment]
i have a [PR](https://github.com/restic/restic/pull/3107) out for this, but i could use some guidance if i were to write a unit test for one specific backend. the test suite seems set up to run a barrage of tests against all the backends, ensuring a basic contract.
